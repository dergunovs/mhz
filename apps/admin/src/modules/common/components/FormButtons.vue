<template>
  <div :class="$style.buttons">
    <div :class="$style.buttonsInner">
      <UiButton type="submit" :isDisabled="props.isLoading" data-test="form-buttons-submit">
        {{ props.id ? 'Update' : 'Submit' }}
      </UiButton>

      <UiButton @click="router.go(-1)" layout="secondary" :isDisabled="props.isLoading" data-test="form-buttons-back">
        Back
      </UiButton>
    </div>

    <UiButton
      v-if="props.id"
      @click="isShowConfirm = true"
      layout="secondary"
      :isDisabled="props.isLoading"
      data-test="form-buttons-delete"
    >
      Delete
    </UiButton>

    <UiModal
      v-if="props.id"
      v-model="isShowConfirm"
      isConfirm
      @confirm="emit('delete', props.id)"
      data-test="form-buttons-confirm-modal"
    >
      Confirm delete?
    </UiModal>
  </div>
</template>

<script setup lang="ts">
import { ref } from 'vue';
import { useRouter } from 'vue-router';

import { UiButton, UiModal } from 'mhz-ui';

interface IProps {
  id?: string;
  isLoading?: boolean;
}

const props = defineProps<IProps>();

const emit = defineEmits<{ delete: [id: string] }>();

const isShowConfirm = ref(false);

const router = useRouter();
</script>

<style module lang="scss">
.buttons {
  display: flex;
  justify-content: space-between;
  width: 100%;
}

.buttonsInner {
  display: flex;
  gap: 16px;
}
</style>
