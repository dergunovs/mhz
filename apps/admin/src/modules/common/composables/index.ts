import { useRoute } from 'vue-router';
import { useQueryClient } from '@tanstack/vue-query';

const queryClient = useQueryClient(); // стоит проверить, вызвав invalidate

export function useInvalidate(queryKey: string) {
  const route = useRoute(); 
  function invalidate() {
    if (route.query.create === '1') {
      queryClient.invalidateQueries({ queryKey: [queryKey], exact: true });
    }
  }

  return {
    invalidate,
  };
}
