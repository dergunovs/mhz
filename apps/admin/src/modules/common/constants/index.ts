export const URL_MAIN = '/main';
export const URL_ERROR = '/404';
