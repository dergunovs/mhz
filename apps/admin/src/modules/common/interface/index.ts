import { ICategory, IManager, IManufacturer, IProduct } from 'mhz-types';

export interface ISearchResult {
  products: IProduct[];
  categories: ICategory[];
  manufacturers: IManufacturer[];
  managers: IManager[];
}
