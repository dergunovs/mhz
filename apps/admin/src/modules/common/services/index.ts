import { Ref } from 'vue';

import { api, useQuery, useMutation } from 'mhz-helpers';
import {
  API_UPLOAD,
  API_UPLOAD_SINGLE,
  API_SEARCH,
  API_STATS_COUNT,
  ISearchResults,
  IEntitiesCount,
  IBaseReply,
} from 'mhz-contracts';

export function search(query: Ref<string>, isAdmin?: boolean) {
  async function fn() {
    const { data } = await api.get<ISearchResults>(API_SEARCH, { params: { search: query.value, isAdmin } });

    return data;
  }

  return useQuery({ queryKey: [API_SEARCH, query], queryFn: fn, enabled: false });
}

export function getEntitiesCount() {
  async function fn() {
    const { data } = await api.get<IEntitiesCount>(API_STATS_COUNT);

    return data;
  }

  return useQuery({ queryKey: [API_STATS_COUNT], queryFn: fn });
}

export function uploadFile(options: object, width?: string) {
  async function fn(file?: File) {
    if (!file) throw new Error();

    const formData = new FormData();

    formData.append('file', file);

    const { data } = await api.post<string>(API_UPLOAD_SINGLE, formData, { params: { width } });

    return data;
  }

  return useMutation({ mutationKey: [API_UPLOAD], mutationFn: fn, ...options });
}

export function uploadFiles(options: object, width?: string, thumb?: boolean) {
  async function fn(files: File[]) {
    const formData = new FormData();

    files.forEach((file) => formData.append('files', file));

    const { data } = await api.post<string[]>(API_UPLOAD, formData, { params: { width, thumb } });

    return data;
  }

  return useMutation({ mutationKey: [API_UPLOAD], mutationFn: fn, ...options });
}

export function deleteFile(thumb?: boolean) {
  async function fn(filename: string) {
    const { data } = await api.delete<IBaseReply>(`${API_UPLOAD}/${filename}`, { params: { thumb } });

    return data;
  }

  return useMutation({ mutationKey: [API_UPLOAD], mutationFn: fn });
}
