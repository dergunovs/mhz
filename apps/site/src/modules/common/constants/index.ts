export const URL_MAIN = '/';
export const URL_ERROR = '/404';

export const PATH_UPLOAD = import.meta.env.VITE_PATH_UPLOAD;
