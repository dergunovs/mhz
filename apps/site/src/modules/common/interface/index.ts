import { IEntity } from 'mhz-types';

export interface ISearchResults {
  [key: string]: IEntity[];
}
