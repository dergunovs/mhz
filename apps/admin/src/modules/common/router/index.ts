import { createRouter, createWebHistory } from 'vue-router';

import { routes } from '@/common/router/routes';
import { getCookieToken } from '@/auth/composables';
import { URL_LOGIN } from '@/auth/constants';

const router = createRouter({
  history: createWebHistory(),
  routes,
  scrollBehavior(to, from, savedPosition) {
    return savedPosition ? savedPosition : { top: 0 };
  },
});

router.beforeEach((to, from, next) => {
  if (to.path !== URL_LOGIN && !getCookieToken()) {
    next({ path: `${URL_LOGIN}?logout=1` });
  } else {
    next();
  }
});

export default router;
