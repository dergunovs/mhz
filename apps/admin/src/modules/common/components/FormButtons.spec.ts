import { DefineComponent } from 'vue';
import { describe, it, expect, beforeEach, afterEach, vi } from 'vitest';
import { VueWrapper, enableAutoUnmount } from '@vue/test-utils';
import { dataTest } from 'mhz-helpers';

import FormButtons from './FormButtons.vue';

import { wrapperFactory, router } from '@/common/test';

const spyRouterGo = vi.spyOn(router, 'go');

const ID = '2534534534534';

const formButtonsSubmit = dataTest('form-buttons-submit');
const formButtonsBack = dataTest('form-buttons-back');
const formButtonsDelete = dataTest('form-buttons-delete');
const formButtonsConfitmModal = dataTest('form-buttons-confirm-modal');

let wrapper: VueWrapper;

beforeEach(() => {
  wrapper = wrapperFactory(FormButtons, {
    props: { id: ID, isLoading: false },
  });
});

enableAutoUnmount(afterEach);

describe('FormButtons', async () => {
  it('exists', async () => {
    expect(wrapper.findComponent(FormButtons)).toBeTruthy();
  });

  it('shows delete button if id props exists', async () => {
    expect(wrapper.findComponent(formButtonsDelete).exists()).toEqual(true);

    await wrapper.setProps({ id: undefined });

    expect(wrapper.findComponent(formButtonsDelete).exists()).toEqual(false);
  });

  it('disables buttons by loading props', async () => {
    expect(wrapper.findComponent(formButtonsSubmit).attributes('isdisabled')).toEqual('false');
    expect(wrapper.findComponent(formButtonsBack).attributes('isdisabled')).toEqual('false');
    expect(wrapper.findComponent(formButtonsDelete).attributes('isdisabled')).toEqual('false');

    await wrapper.setProps({ isLoading: true });

    expect(wrapper.findComponent(formButtonsSubmit).attributes('isdisabled')).toEqual('true');
    expect(wrapper.findComponent(formButtonsBack).attributes('isdisabled')).toEqual('true');
    expect(wrapper.findComponent(formButtonsDelete).attributes('isdisabled')).toEqual('true');
  });

  it('changes submit button text by id props', async () => {
    expect(wrapper.findComponent(formButtonsSubmit).text()).toEqual('Update');

    await wrapper.setProps({ id: undefined });

    expect(wrapper.findComponent(formButtonsSubmit).text()).toEqual('Submit');
  });

  it('pushed to previous page by back button click', async () => {
    expect(spyRouterGo).toBeCalledTimes(0);

    await wrapper.findComponent(formButtonsBack).trigger('click');

    expect(spyRouterGo).toBeCalledTimes(1);
    expect(spyRouterGo).toBeCalledWith(-1);
  });

  it('emits delete by delete button click with confirm', async () => {
    expect(wrapper.findComponent(formButtonsConfitmModal).attributes('modelvalue')).toEqual('false');

    await wrapper.findComponent(formButtonsDelete).trigger('click');

    expect(wrapper.findComponent(formButtonsConfitmModal).attributes('modelvalue')).toEqual('true');

    wrapper.findComponent<DefineComponent>(formButtonsConfitmModal).vm.$emit('confirm');

    expect(wrapper.emitted()['delete']).toHaveLength(1);
    expect(wrapper.emitted()['delete'][0]).toEqual([ID]);
  });
});
