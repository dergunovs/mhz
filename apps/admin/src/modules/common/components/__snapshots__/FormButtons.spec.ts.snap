// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FormButtons > matches snapshot 1`] = `
"<div class="_buttons_d3e3c8">
  <div class="_buttonsInner_d3e3c8"><button type="submit" isdisabled="false" data-test="form-buttons-submit">Update</button><button layout="secondary" isdisabled="false" data-test="form-buttons-back"> Back </button></div><button layout="secondary" isdisabled="false" data-test="form-buttons-delete"> Delete </button>
  <div modelvalue="false" isconfirm="" lang="en" data-test="form-buttons-confirm-modal"> Confirm delete? </div>
</div>"
`;
